### PR TITLE
modelTypes: Sync Message type with the doc, at FL 132

### PR DESCRIPTION
**Pull request by chrisbobbe**

### CODE REVIEW

**Comment by chrisbobbe** [comment]
Looks like a flake in some new Jest tests:

```
Summary of all failing tests
FAIL src/__tests__/permissionSelectors-test.js
  ● getCanCreatePublicStreams › returns true when policy is 3; role is 400; waitingPeriodPassed is true
    expect(received).toBe(expected) // Object.is equality
    Expected: true
    Received: false
      132 |       }
      133 |
    > 134 |       expect(getCanCreatePublicStreams(newState)).toBe(expected);
          |                                                                  ^
      135 |     },
      136 |   );
      137 | });
      at _callee$ (src/__tests__/permissionSelectors-test.js:134:83)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:294:22)
      at Generator.next (node_modules/regenerator-runtime/runtime.js:119:21)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at invoke (node_modules/regenerator-runtime/runtime.js:155:20)
      at node_modules/regenerator-runtime/runtime.js:190:11
      at callInvokeWithMethodAndArg (node_modules/regenerator-runtime/runtime.js:189:16)
      at AsyncIterator.enqueue (node_modules/regenerator-runtime/runtime.js:212:13)
  ● getCanCreatePrivateStreams › returns true when policy is 3; role is [400](https://github.com/zulip/zulip-mobile/runs/6767338490?check_suite_focus=true#step:7:401); waitingPeriodPassed is true
    expect(received).toBe(expected) // Object.is equality
    Expected: true
    Received: false
      219 |       }
      220 |
    > 221 |       expect(getCanCreatePrivateStreams(newState)).toBe(expected);
          |                                                                   ^
      222 |     },
      223 |   );
      224 | });
      at _callee2$ (src/__tests__/permissionSelectors-test.js:221:84)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:294:22)
      at Generator.next (node_modules/regenerator-runtime/runtime.js:119:21)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at invoke (node_modules/regenerator-runtime/runtime.js:155:20)
      at node_modules/regenerator-runtime/runtime.js:190:11
      at callInvokeWithMethodAndArg (node_modules/regenerator-runtime/runtime.js:189:16)
      at AsyncIterator.enqueue (node_modules/regenerator-runtime/runtime.js:212:13)
Test Suites: 1 failed, 195 passed, 196 total
Tests:       2 failed, 4 skipped, 2 todo, 4729 passed, 4737 total
Snapshots:   194 passed, 194 total
Time:        152.832 s
Ran all test suites in 2 projects.
```

**Comment by chrisbobbe** [comment]
For the flake: #5402

**Comment by chrisbobbe** [comment]
(pushed again to trigger CI to run again)

File: src/api/modelTypes.js [hunk 1]
```diff
@@ -655,63 +674,76 @@ export type Submessage = $ReadOnly<{|
...
+  //
+  // Properties on `event.message` in `message` events. Order follows the
+  // doc. See comment on Message for how up-to-date these are.
+  //
 
   client: string,
   content: string,
   content_type: 'text/html',
   edit_history: $ReadOnlyArray<MessageEdit>,
-  gravatar_hash: string,
```
Comment (gnprice) [inline src/api/modelTypes.js:696]: modelTypes: Remove gravatar_hash from Message, deprecated in 2013 (really?)
    
    See zulip/zulip@9f7236eec, in 2017-05, which says this had been
    "deprecated and replaced" four years previously.

:smile: 

```
commit 9f7236eec1c16fe8333824612520b410d3a00046
Author: Tim Abbott <tabbott@zulip.com>
Date:   Tue May 9 22:20:47 2017 -0700

    message: Remove unused old gravatar_hash field from message dicts.
    
    This was deprecated and replaced some 4 years ago.
…
-            gravatar_hash     = gravatar_hash(sender_email), # Deprecated June 2013
             avatar_url        = avatar_url,
```

And following with `git log -G gravatar_hash` finds: zulip/zulip@d740d7c08274859b43a10899dd1f4cd45df76f31, which introduced `avatar_url`, with summary line "Show user-uploaded avatars on the website."

So yeah, apparently it was deprecated with the introduction of avatars that *weren't* from Gravatar and instead were avatars you uploaded to Zulip itself.

(Being originally gravatar-only sounds a bit less unreasonable when you realize that Gravatar isn't just about autogenerated avatars -- it's meant to be a site where you upload an avatar and then have it work across the Web.)

File: src/api/modelTypes.js [hunk 2]
```diff
@@ -655,63 +674,76 @@ export type Submessage = $ReadOnly<{|
...
-
-  /**
-   * Present under EVENT_NEW_MESSAGE and under MESSAGE_FETCH_COMPLETE,
-   * and in `state.messages`, all as an AvatarURL, because we
-   * translate into that form at the edge.
-   *
-   * For how it appears at the edge (and how we translate) see
-   * AvatarURL.fromUserOrBotData.
-   */
-  avatar_url: AvatarURL,
```
Comment (gnprice) [inline src/api/modelTypes.js:688]: modelTypes [nfc]: Reorder MessageBase type's properties to follow the doc
    
    And give a clear section to the ones that aren't present in
    `message` events. (The jsdoc suggests we should primarily model the
    Message type on `message` events.)

Good idea!

`avatar_url` should stay at the top of the list, right? I see it at the top here: https://zulip.com/api/get-events#message

Comment (gnprice) [reply src/api/modelTypes.js:688]: (It has a different type in the actual event, but is present, and it's another form of the same data.)

File: src/api/modelTypes.js [hunk 3]
```diff
@@ -655,63 +674,76 @@ export type Submessage = $ReadOnly<{|
...
-  // The rest are believed to really appear in `message` events.
+  //
+  // Properties on `event.message` in `message` events. Order follows the
+  // doc. See comment on Message for how up-to-date these are.
+  //
 
   client: string,
   content: string,
   content_type: 'text/html',
   edit_history: $ReadOnlyArray<MessageEdit>,
```
Comment (gnprice) [inline src/api/modelTypes.js:685]: While reordering to match the doc, I think it'd be helpful to also insert mentions here of the properties that do appear in the doc and aren't included in this type:
```suggestion
  content_type: 'text/html',
  // display_recipient handled on PmMessage and StreamMessage separately
  edit_history: $ReadOnlyArray<MessageEdit>,
```

Similarly later
```
  // stream_id handled on StreamMessage
```
and some others.

That just further helps make it easy to compare the type definition to the doc.

File: src/api/modelTypes.js [hunk 4]
```diff
@@ -655,63 +674,76 @@ export type Submessage = $ReadOnly<{|
...
-   * For how it appears at the edge (and how we translate) see
-   * AvatarURL.fromUserOrBotData.
-   */
-  avatar_url: AvatarURL,
-
-  // The rest are believed to really appear in `message` events.
+  //
+  // Properties on `event.message` in `message` events. Order follows the
+  // doc. See comment on Message for how up-to-date these are.
+  //
```
Comment (gnprice) [inline src/api/modelTypes.js:680]: Let's include a doc link, too: https://zulip.com/api/get-events#message

**Review by gnprice — COMMENTED** [review]
Thanks! This looks great; small comments below, then please merge at will.

**Comment by chrisbobbe** [comment]
Thanks for the review!

> small comments below, then please merge at will.

Done, after addressing the comments.

**Pull request merged**